### PR TITLE
Test DeepEnsemble with float32

### PR DESCRIPTION
**Pull request by uri-granta**
**Related issue(s)/PRs:** 

## Summary

...

**Fully backwards compatible:** yes / no

## PR checklist

- [ ] The quality checks are all passing
- [ ] The bug case / new feature is covered by tests
- [ ] Any new features are well-documented (in docstrings or notebooks)

### CODE REVIEW

File: tests/integration/test_bayesian_optimization.py [hunk 1]
```diff
@@ -603,12 +622,21 @@ def _test_optimizer_finds_minimum(
...
+            tf.cast(search_space.upper, dtype=tf.float32),
+        )
+
     if model_type in [SparseVariational, DeepEnsemble]:
         num_initial_query_points = 20
     elif model_type in [DeepGaussianProcess]:
         num_initial_query_points = 25
 
     initial_query_points = search_space.sample(num_initial_query_points)
+    assert initial_query_points.dtype is tf.float32 if single_precision else tf.float64
```
Comment (hstojic) [inline tests/integration/test_bayesian_optimization.py]: maybe also test whether generated query points are of correct dtype?

File: tests/integration/test_bayesian_optimization.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
 import tempfile
 from functools import partial
 from pathlib import Path
-from typing import Any, List, Mapping, Optional, Tuple, Type, Union, cast
+from typing import Any, Callable, List, Mapping, Optional, Tuple, Type, Union, cast
```
Comment (hstojic) [inline tests/integration/test_bayesian_optimization.py:21]: we should do the same with ask-tell, no?

**Review by hstojic — COMMENTED** [review]
a step in the right direction :)
left some comments - this checks if all runs, which is necessary, but certain parts of the pipeline might nevertheless cast back and forth which we don't test here - I'm not sure whats the best way to do that, just saying this is not sufficient - we'll probably need to  test each individual component, or perhaps each individual object should have a dtype property which could facilitate that...

cannot recall now, but do we have dtype checks in unit tests of the DE model - e.g. checking prediction, sample methods, and trajectory sampler?

File: tests/integration/test_bayesian_optimization.py (hunk 2 above)
Comment (uri-granta) [reply tests/integration/test_bayesian_optimization.py:21]: Yes, though I'll postpone that to a future PR as at the moment we don't have any DeepEnsemble tests with ask-tell. This PR was basically an initial nothing-blows-up check. Next step will be to (1) figure out how to check whether any conversions are happening under the hood (2) write lots more tests.

**Comment by uri-granta** [comment]
By patching tf.cast I've verified that there are no unexpected casts happening in the test. The only casts are in `sample_model_index` (which casts deep ensemble size to tf.int32) and `write_summary_data_based_metrics` (which casts  dataset observations to the model prediction dtype).

**Review by hstojic — APPROVED** [review]
the diff became a bit mangled, but I think its good now

**Comment by uri-granta** [comment]
> the diff became a bit mangled, but I think its good now

FYI you can add ?w=1 to the diff url to get it to ignore whitespace, which makes it much more readable!

**Pull request merged**